tests/integration/replica-redirect.tcl

```diff
@@ -93,12 +98,154 @@ start_server {tags {needs:repl external:skip}} {
+            $rd_wait read ; # Consume the OK reply
+            $rd_wait set foo bar
+            $rd_wait read
+            $rd_wait wait 1 0 ; # Blocks as we can't sync
+
+            # XREAD is a reading command and thus, should
+            # not be redirected if the client is read only.
+            set rd_xread [valkey_deferring_client -1]
+            $rd_xread client capa redirect
+            assert_match "OK" [$rd_xread read]
+            $rd_xread readonly
+            assert_match "OK" [$rd_xread read]
+            $rd_xread xread block 0 streams k 0
+
+            wait_for_blocked_clients_count 3 100 10 -1
+
+            r -1 failover
+
+            # The replica is not synced and sleeps. Primary is waiting for
+            # sync.
+            assert_equal "waiting-for-sync" [s -1 master_failover_state]
+
+            set rd_brpop_after [valkey_deferring_client -1]
+            $rd_brpop_after client capa redirect
+            assert_match "OK" [$rd_brpop_after read]
+            $rd_brpop_after brpop list 0
+
+            wait_for_blocked_clients_count 4 100 10 -1
+
+            resume_process [srv 0 pid]
+
+            # Wait for failover to end
+            wait_for_condition 50 100 {
+                [s -1 master_failover_state] == "no-failover"
+            } else {
+                fail "primary not in no-failover state"
+            }
 
             assert_match *master* [r role]
             assert_match *slave* [r -1 role]
 
+            # REDIRECT response shall be sent when the new primary is ready,
+            # i.e. when unblocking all clients after failover
+            assert_error "REDIRECT $replica_host:$replica_port" {$rd_brpop_before read}
+            assert_error "REDIRECT $replica_host:$replica_port" {$rd_brpop_after read}
+            assert_error "REDIRECT $replica_host:$replica_port" {$rd_xread read}
+
+            # WAIT is unblocked in the main loop. Thus, it must have succeeded by now.
+            assert_equal "1" [$rd_wait read]
+
+            $rd_brpop_before close
+            $rd_brpop_after close
+            $rd_wait close
+            $rd_xread close
+        }
+    }
+}
+
+start_server {tags {needs:repl external:skip}} {
+    start_server {} {
+        set master [srv 0 client]
+        set master_host [srv 0 host]
+        set master_port [srv 0 port]
+        set replica [srv -1 client]
+        set replica_host [srv -1 host]
+        set replica_port [srv -1 port]
+
+        $replica replicaof $master_host $master_port
+        wait_for_condition 100 100 {
+            [string match *slave*online* [$master info replication]] &&
+            [s -1 master_link_status] eq {up}
+        } else {
+            fail "Replica not online."
+        }
+
+        test {client with REDIRECT capability receives REDIRECT error during failover} {
+            set rd [valkey_deferring_client]
+            $rd client capa redirect
+            assert_match "OK" [$rd read]
+
```

Comment:
Perhaps we should verify the `XREAD` case for a ` READONLY` client as well here?